Mac OS X build with Python 3.6 fails

One representative error:

```
torch/csrc/autograd/functions/init.cpp:220:37: error: address of overloaded function 'getTupleAttr' does not match required type '_object *(_object *, void *)'
  {(char*)"output_padding", (getter)getTupleAttr<ConvBackwardBackward, std::vector<int>, ConvParams,
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
torch/csrc/autograd/functions/init.cpp:82:11: note: candidate template ignored: invalid explicitly-specified argument for template parameter 'Convert'
PyObject* getTupleAttr(PyObject* obj, void* _unused)
```

The cause of the problem is aa911939a328eff55c9b28b39ed3c43507ba8a2a:

```
   {(char*)"output_padding", (getter)getTupleAttr<ConvForward, std::vector<int>, ConvParams,
-                                         &ConvParams::output_padding, long, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::output_padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
```

It seems that on clang, changing the type parameter here is sufficient to cause template instantiation to fail.

Maybe the easiest way to fix this is to write a more portable version of PyInt_FromLong (and friends) which always returns `int64_t`.

--- a/torch/csrc/autograd/functions/init.cpp
+++ b/torch/csrc/autograd/functions/init.cpp
@@ -122,6 +122,15 @@ PyObject* getTensorAttr(PyObject* obj, void* _unused)
   END_HANDLE_TH_ERRORS
 }
 
+// The type for PyInt_FromLong varies between Windows (int64_t) and
+// Linux (long).  Unfortunately, on Clang, if the template type argument
+// does not *exactly* line up, Clang discards the template
+// specialization.  So we write a little stub function which does have
+// the right type.
+static PyObject* PortablePyInt_FromLong(int64_t ival) {
+  return PyInt_FromLong(ival);
+}
+
 static struct PyGetSetDef batch_norm_forward_properties[] = {
   THP_FUNCTION_DEFAULT_PROPERTIES,
   {(char*)"running_mean", (getter)getTensorAttr<BatchNormForward, BatchNormParams,
@@ -176,51 +185,51 @@ static struct PyGetSetDef batch_norm_backward_backward_properties[] = {
 static struct PyGetSetDef conv_forward_properties[] = {
   THP_FUNCTION_DEFAULT_PROPERTIES,
   {(char*)"stride", (getter)getTupleAttr<ConvForward, std::vector<int>, ConvParams,
-                                         &ConvParams::stride, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::stride, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"padding", (getter)getTupleAttr<ConvForward, std::vector<int>, ConvParams,
-                                         &ConvParams::padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::padding, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"dilation", (getter)getTupleAttr<ConvForward, std::vector<int>, ConvParams,
-                                         &ConvParams::dilation, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::dilation, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"transposed", (getter)getValueAttr<ConvForward, bool, ConvParams,
                                          &ConvParams::transposed, long, PyBool_FromLong>, NULL, NULL, NULL},
   {(char*)"output_padding", (getter)getTupleAttr<ConvForward, std::vector<int>, ConvParams,
-                                         &ConvParams::output_padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::output_padding, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"groups", (getter)getValueAttr<ConvForward, int, ConvParams,
-                                         &ConvParams::groups, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::groups, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {NULL}
 };
 
 static struct PyGetSetDef conv_backward_properties[] = {
   THP_FUNCTION_DEFAULT_PROPERTIES,
   {(char*)"stride", (getter)getTupleAttr<ConvBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::stride, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::stride, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"padding", (getter)getTupleAttr<ConvBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::padding, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"dilation", (getter)getTupleAttr<ConvBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::dilation, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::dilation, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"transposed", (getter)getValueAttr<ConvBackward, bool, ConvParams,
                                          &ConvParams::transposed, long, PyBool_FromLong>, NULL, NULL, NULL},
   {(char*)"output_padding", (getter)getTupleAttr<ConvBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::output_padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::output_padding, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"groups", (getter)getValueAttr<ConvBackward, int, ConvParams,
-                                         &ConvParams::groups, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::groups, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {NULL}
 };
 
 static struct PyGetSetDef conv_backward_backward_properties[] = {
   THP_FUNCTION_DEFAULT_PROPERTIES,
   {(char*)"stride", (getter)getTupleAttr<ConvBackwardBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::stride, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::stride, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"padding", (getter)getTupleAttr<ConvBackwardBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::padding, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"dilation", (getter)getTupleAttr<ConvBackwardBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::dilation, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::dilation, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"transposed", (getter)getValueAttr<ConvBackwardBackward, bool, ConvParams,
                                          &ConvParams::transposed, long, PyBool_FromLong>, NULL, NULL, NULL},
   {(char*)"output_padding", (getter)getTupleAttr<ConvBackwardBackward, std::vector<int>, ConvParams,
-                                         &ConvParams::output_padding, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::output_padding, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {(char*)"groups", (getter)getValueAttr<ConvBackwardBackward, int, ConvParams,
-                                         &ConvParams::groups, int64_t, PyInt_FromLong>, NULL, NULL, NULL},
+                                         &ConvParams::groups, int64_t, PortablePyInt_FromLong>, NULL, NULL, NULL},
   {NULL}
 };
 

